Crash: Not reached: unreported empty/missing binding: `MissingBindingImpl(<...>)`

Full stack trace:
```
java.lang.AssertionError: Not reached: unreported empty/missing binding: `MissingBindingImpl(target=com.yandex.yatagan.core.model.impl.NodeModelImpl@46a7b47, owner=com.yandex.yatagan.core.graph.impl.BindingGraphImpl@6d340d36)`
	at com.yandex.yatagan.codegen.impl.CreationGeneratorVisitor.visitEmpty(generateCreation.kt:268)
	at com.yandex.yatagan.codegen.impl.CreationGeneratorVisitor.visitEmpty(generateCreation.kt:44)
	at com.yandex.yatagan.core.graph.impl.bindings.MissingBindingImpl.accept(MissingBindingImpl.kt:50)
	at com.yandex.yatagan.codegen.impl.GenerateCreationKt.generateCreation(generateCreation.kt:285)
	at com.yandex.yatagan.codegen.impl.InlineCreationStrategy.generateAccess(accessStrategies.kt:175)
	at com.yandex.yatagan.codegen.impl.GeneratorUtilsKt.generateAccess(generatorUtils.kt:68)
	at com.yandex.yatagan.codegen.impl.CreationGeneratorVisitor$visitProvision$1$1.genArgs(generateCreation.kt:60)
	at com.yandex.yatagan.codegen.impl.CreationGeneratorVisitor$visitProvision$1$1.visitConstructor(generateCreation.kt:92)
	at com.yandex.yatagan.codegen.impl.CreationGeneratorVisitor$visitProvision$1$1.visitConstructor(generateCreation.kt:57)
	at com.yandex.yatagan.lang.common.ConstructorBase.accept(ConstructorBase.kt:24)
	at com.yandex.yatagan.codegen.impl.CreationGeneratorVisitor.visitProvision(generateCreation.kt:57)
	at com.yandex.yatagan.codegen.impl.CreationGeneratorVisitor.visitProvision(generateCreation.kt:44)
	at com.yandex.yatagan.core.graph.impl.bindings.InjectConstructorProvisionBindingImpl.accept(InjectConstructorProvisionBindingImpl.kt:61)
	at com.yandex.yatagan.codegen.impl.GenerateCreationKt.generateCreation(generateCreation.kt:285)
	at com.yandex.yatagan.codegen.impl.InlineCreationStrategy.generateAccess(accessStrategies.kt:175)
	at com.yandex.yatagan.codegen.impl.GeneratorUtilsKt.generateAccess(generatorUtils.kt:68)
	at com.yandex.yatagan.codegen.impl.ComponentGenerator.generate(ComponentGenerator.kt:89)
	at com.yandex.yatagan.codegen.impl.ComponentGenerator.generate(ComponentGenerator.kt:166)
	at com.yandex.yatagan.codegen.impl.ComponentGeneratorFacade.generateTo(ComponentGeneratorFacade.kt:47)
	at com.yandex.yatagan.processor.common.ProcessKt.process(process.kt:104)
	at com.yandex.yatagan.processor.ksp.KspYataganProcessor.process(KspYataganProcessor.kt:47)
        ...
```

Code for reproduction:
```kotlin
import com.yandex.yatagan.*
import javax.inject.*
interface Api
interface Api2
class Impl @Inject constructor(api2: Api2): Api
class Impl2 @Inject constructor(): Api2

@Module interface MyModule1 {
    @Binds fun api(i: Impl): Api
    @Binds fun api2(i: Impl2): Api2
}
@Module interface MyModule2 {
    @Binds fun api(i: Impl): Api
}

@Component interface RootComponent {
    fun createSub2(): SubComponent2
    fun createSub1(): SubComponent1
}

@Component(isRoot = false, modules = [MyModule1::class]) interface SubComponent1 {
    val api: Api
}
@Component(isRoot = false, modules = [MyModule2::class]) interface SubComponent2 {
    val api: Api
}
```
The problem is, that specific `equals`/`hashCode` for `AliasBindingImpl` was written to avoid reporting equivalent `@Binds` as duplicates (was done in #61). But the validation pipeline puts visited nodes to hash sets. And in this specific case equivalent (but physically different) aliases may end up hiding some graph errors, as validiator thinks that is already visited the alias, but in reaility it visited its "twin". This bug may lead to incorrect graph being generated or crash the generator.

--- a/core/graph/impl/src/main/kotlin/GraphBindingsManager.kt
+++ b/core/graph/impl/src/main/kotlin/GraphBindingsManager.kt
@@ -336,7 +336,9 @@ internal class GraphBindingsManager(
             }
 
             if (bindings.size > 1) {
-                val distinct = bindings.toSet()
+                val distinct = bindings
+                        .map { it.maybeUnwrapSyntheticAlias() }
+                        .distinctBy { it.accept(AliasEquivalenceSelector) }
                 if (distinct.size > 1) {
 
                     // We tolerate multibinding duplicates, because of the "extends" behavior.
@@ -478,6 +480,14 @@ internal class GraphBindingsManager(
         val keyType: Type,
         val valueType: NodeModel,
     )
+
+    private object AliasEquivalenceSelector : BaseBinding.Visitor<Any> {
+        override fun visitAlias(alias: AliasBinding): Pair<NodeModel, NodeModel> {
+            // We compare aliases only as a pair {source; target} in terms of duplicate check.
+            return alias.source to alias.target
+        }
+        override fun visitBinding(binding: Binding) = binding
+    }
 
     companion object Key : Extensible.Key<GraphBindingsManager> {
         override val keyType get() = GraphBindingsManager::class.java

--- a/core/graph/impl/src/main/kotlin/bindings/AliasBindingImpl.kt
+++ b/core/graph/impl/src/main/kotlin/bindings/AliasBindingImpl.kt
@@ -39,17 +39,6 @@ internal class AliasBindingImpl(
 
     override val source get() = impl.sources.single()
 
-    override fun equals(other: Any?): Boolean {
-        return this === other || (other is AliasBinding &&
-                source == other.source && target == other.target)
-    }
-
-    override fun hashCode(): Int {
-        var result = target.hashCode()
-        result = 31 * result + source.hashCode()
-        return result
-    }
-
     override fun toString(childContext: MayBeInvalid?) = bindingModelRepresentation(
         modelClassName = "alias",
         childContext = childContext,

--- a/testing/tests/src/main/kotlin/CompileTestDriverBase.kt
+++ b/testing/tests/src/main/kotlin/CompileTestDriverBase.kt
@@ -163,24 +163,30 @@ abstract class CompileTestDriverBase private constructor(
                 System.err.println(messageLog)
                 Assert.assertTrue("Compilation failed, yet expected output is blank", goldenOutput.isNotBlank())
             }
+
+            generatedFilesSubDir()?.let {
+                val goldenFiles = GoldenSourceRegex.findAll(
+                        javaClass.getResourceAsStream("/$goldenCodeResourcePath")
+                                ?.bufferedReader()?.readText() ?: ""
+                ).associateByTo(
+                        destination = mutableMapOf(),
+                        keySelector = { it.groupValues[1] },
+                        valueTransform = { it.groupValues[2].trim() },
+                )
+
+                for (generatedFile in generatedFiles) {
+                    val goldenContents = goldenFiles.remove(generatedFile.relativePath) ?: "<unexpected file>"
+                    Assert.assertEquals("Generated file '${generatedFile.relativePath}' doesn't match the golden",
+                            goldenContents, generatedFile.contents.trim())
+                }
+            }
         } finally {
             generatedFilesSubDir()?.let { generatedFilesSubDir ->
-                val goldenFiles = GoldenSourceRegex.findAll(
-                    javaClass.getResourceAsStream("/$goldenCodeResourcePath")
-                        ?.bufferedReader()?.readText() ?: ""
-                ).associateByTo(
-                    destination = mutableMapOf(),
-                    keySelector = { it.groupValues[1] },
-                    valueTransform = { it.groupValues[2].trim() },
-                )
                 for (generatedFile in generatedFiles) {
                     val fileUrl = "file://" + workingDir
                         .resolve(generatedFilesSubDir)
                         .resolve(generatedFile.relativePath).absolutePath
                     println("Generated $fileUrl")
-                    val goldenContents = goldenFiles.remove(generatedFile.relativePath) ?: "<unexpected file>"
-                    Assert.assertEquals("Generated file '${generatedFile.relativePath}' doesn't match the golden",
-                        goldenContents, generatedFile.contents.trim())
                 }
             }
         }

--- a/testing/tests/src/test/kotlin/CoreBindingsFailureTest.kt
+++ b/testing/tests/src/test/kotlin/CoreBindingsFailureTest.kt
@@ -909,4 +909,38 @@ class CoreBindingsFailureTest(
 
         compileRunAndValidate()
     }
+
+    @Test
+    fun `case for #88`() {
+        givenKotlinSource("test.TestCase", """
+            import com.yandex.yatagan.*
+            import javax.inject.*
+            interface Api
+            interface Api2
+            class Impl @Inject constructor(api2: Api2): Api
+            class Impl2 @Inject constructor(): Api2
+
+            @Module interface MyModule1 {
+                @Binds fun api(i: Impl): Api
+                @Binds fun api2(i: Impl2): Api2
+            }
+            @Module interface MyModule2 {
+                @Binds fun api(i: Impl): Api
+            }
+
+            @Component interface RootComponent {
+                fun createSub2(): SubComponent2
+                fun createSub1(): SubComponent1
+            }
+
+            @Component(isRoot = false, modules = [MyModule1::class]) interface SubComponent1 {
+                val api: Api
+            }
+            @Component(isRoot = false, modules = [MyModule2::class]) interface SubComponent2 {
+                val api: Api
+            }
+        """.trimIndent())
+
+        compileRunAndValidate()
+    }
 }

--- a/testing/tests/src/test/resources/golden/CoreBindingsFailureTest/case-for-88.golden.txt
+++ b/testing/tests/src/test/resources/golden/CoreBindingsFailureTest/case-for-88.golden.txt
@@ -1,0 +1,13 @@
+error: Missing binding for test.Api2
+NOTE: No known way to infer the binding
+Encountered:
+  in graph for root-component test.RootComponent
+  in graph for component test.SubComponent2
+  in entry-point getApi: test.Api
+                         ^-[*1]--
+  in [1*] alias test.MyModule2::api(test.Impl)
+                                    ^-[*2]---
+  in [2*] inject-constructor test.Impl(test.Api2)
+                                       ^-[*3]---
+  here: [3*] <missing>
+             ^~~~~~~~~

